src/v/cluster/topics_frontend.cc

```diff
@@ -949,4 +982,137 @@ topics_frontend::do_cancel_moving_partition_replicas(
     co_return results;
 }
 
+ss::future<std::error_code> topics_frontend::change_replication_factor(
+  model::topic_namespace topic,
+  cluster::replication_factor_type new_replication_factor,
+  model::timeout_clock::time_point timeout) {
+    auto tp_metadata = _topics.local().get_topic_metadata_ref(topic);
+    if (!tp_metadata.has_value()) {
+        co_return errc::topic_not_exists;
+    }
+    if (!tp_metadata.value().get().is_topic_replicable()) {
+        co_return errc::topic_operation_error;
+    }
+
+    auto current_replication_factor
+      = tp_metadata.value().get().get_replication_factor();
+
+    if (current_replication_factor == new_replication_factor) {
+        co_return errc::success;
+    }
+
+    new_replicas_assigment new_replicas;
+
+    // We should hold allocation_units from reallocate_partition while we do
+    // replicate_and_wait.
+    std::optional<allocation_units> allocation_units;
```

Comment:
I would make `replicate_and_wait` to be called from inside of `increase_replication_factor` and `decrease_replication_factor` this way you wouldn't have to create those local variables upfront and take units by reference. WDYT ?